File: algorithms/2_medium/P/01104_path_in_zigzag_labelled_binary_tree/solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

static inline int calc_pow_2(int n)
{
    int res = 0;
    while (n > 0) {
        res += 1;
        n /= 2;
    }
    return res;
}

int* pathInZigZagTree(const int label, int* return_len)
{
    assert(label >= 1 && return_len != NULL);
    const int pow_2 = calc_pow_2(label);
    int* res = malloc((size_t)pow_2 * sizeof(*res));
    assert(res != NULL);
    *return_len = pow_2;
    int curr_val = label;
    int hi = (1 << pow_2) - 1;
    int lo = 1 << (pow_2 - 1);

    for (int i = pow_2 - 1; i >= 0; i -= 1) {
        res[i] = curr_val;

        curr_val = (hi + lo - curr_val) / 2;
        hi /= 2;
        lo /=2;
    }

    return res;
}

File: algorithms/2_medium/P/01104_path_in_zigzag_labelled_binary_tree/solution.rs
impl Solution {
    #[inline(always)]
    fn calc_pow_2(mut n: i32) -> i32 {
        let mut res = 0;
        while n > 0 {
            res += 1;
            n /= 2;
        }
        res
    }

    pub fn path_in_zig_zag_tree(label: i32) -> Vec<i32> {
        assert!(label >= 1);
        let pow_2 = Self::calc_pow_2(label);
        let mut res = vec![0_i32; pow_2 as usize];
        let mut curr_val = label;
        let mut hi = 2_i32.pow(pow_2 as u32) - 1;
        let mut lo = 2_i32.pow((pow_2 - 1) as u32);

        for i in (0..pow_2 as usize).rev() {
            res[i] = curr_val;

            curr_val = (hi + lo - curr_val) / 2;
            hi /= 2;
            lo /=2;
        }

        res
    }
}
